connection-drain-middleware.js

```diff
@@ -0,0 +1,167 @@
+'use strict';
+const util = require('util');
+const LRU = require('lru-cache');
+const PromiseTool = require('./promise-native-tool');
+const Exception = require('./exception');
+
+const INIT_ERROR_CODE = 'INIT_ERROR_CODE';
+
+/**
+ * This is to provide a smooth mechanism for server shutdown and connection draining by:
+ *      * Add "Connection: close" every X seconds to prevent infinite keep-alive connections
+ *      * Handle SIGTERM signal marking the app as draining and shutting down the http server when received
+ *      * Provide a health-check endpoint (by default in /_health) that returns HTTP 200 normally or HTTP 500 when the app is draining
+ *
+ * Usage:
+ *
+ *      const ConnectionDrainMiddlewarw = require('sq-toolkit/connection-drain-middleware');
+ *
+ *  - on app.init():
+ *      ...
+ *      this.app = express();
+ *      this.app.use(requestMiddleware);  // this is optional
+ *      this.app.locals.connectionDrainHelper = new ConnectionDrainingHelper(app, logger);
+ *
+ *  - on app.start():
+ *      ...
+ *      this._server = this.app.listen(port);
+ *      this.app.locals.connectionDrainHelper.setHttpServer(this._server);
+ */
+class ConnectionDrainMiddleware {
+    /**
+     * Setup health check and connection draining mechanism to work with kubernetes/load balancing infrastructure
+     * @param anExpressApp - The express application instance
+     * @param logger - A sq-logger instance
+     * @param {Object=} options
+     * @param {number=40} options.keepAliveBreakSeconds - Interval after which keep-alive connections will be break
+     * @param {number=5} options.keepAliveBreakDeltaSeconds - Limit for random secs added/subtracted to the break interval of each connection to prevent breaking every connection at the same moment
+     * @param {number=10} options.httpServerCloseTimeoutSeconds - Timeout to wait the http server for graceful shutdown once SIGTERM is received. After that the process will exit
+     * @param {string='connection-drain-helper'} options.logPrefix - Prefix to add to every log line within this middleware
+     * @param {string='/_health'} options.healthCheckEndpoint - Endpoint for the health-check controller
+     */
+    constructor(anExpressApp, logger, options = {}) {
+        if(anExpressApp == null){
+            throw new Exception(INIT_ERROR_CODE, 'A express app must be provided.');
+        }
+        if(logger == null){
+            throw new Exception(INIT_ERROR_CODE, 'A sq-logger instance must be provided.');
+        }
+        this.expressApp = anExpressApp;
+        this.fdMap = new LRU({ max: 10000 });
+        this.logger = logger;
+        this.keepAliveBreakSeconds = options.keepAliveBreakSeconds || 40;
+        this.keepAliveBreakDeltaSeconds = options.keepAliveBreakDeltaSeconds || 5;
+        this.httpServerCloseTimeoutSeconds = options.httpServerCloseTimeoutSeconds || 10;
```

Comment:
This paras should be check against != null instead, because 0 cannot be passed
`options.httpServerCloseTimeoutSeconds != null ? options.httpServerCloseTimeoutSeconds : 10`